lib/meta_logger/formatter.ex

```diff
@@ -0,0 +1,98 @@
+defprotocol MetaLogger.Formatter do
+  @doc """
+  MetaLogger forrmatter protocol
+  """
+  @fallback_to_any true
+
+  @spec format(struct()) :: String.t() | list()
+  def format(payload)
+
+  defmodule IncorrectPayload do
+    defexception [:message]
+  end
+
+  defmodule IncorrectOrNotSetFormatterFunction do
+    defexception [:message]
+  end
+end
+
+defimpl MetaLogger.Formatter, for: Any do
+  @moduledoc """
+  Default implementation for any struct which used to be derived.
+  """
+  defmacro __deriving__(module, _struct, options) do
+    formatter_func = fetch_formatter_function(options)
+
+    quote do
+      defimpl MetaLogger.Formatter, for: unquote(module) do
+        @replacement "[FILTERED]"
+
+        def format(struct) do
+          formatter_fn = Keyword.fetch!(unquote(options), :formatter_fn)
+
+          struct
+          |> Map.get(:payload)
+          |> safe_func_invoke()
+          |> filter()
+        end
+
+        defp filter(payload) when is_list(payload),
+          do: Enum.map(payload, &filter(&1))
+
+        defp filter(payload) when is_bitstring(payload) do
+          Keyword.get(unquote(options), :filter_patterns)
+          |> filter(payload)
+        end
+
+        defp filter(patterns, payload) do
+          Enum.reduce(patterns, payload, fn pattern, payload ->
+            String.replace(payload, ~r/#{pattern}/, @replacement)
```

Comment:
can we define the patterns at compile time using module attribute?

Reply:
Could be. I need to think how. This metaprogramming still confuses me a bit :)